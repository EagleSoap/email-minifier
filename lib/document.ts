export const createDocument = async (html: string): Promise<Document> => {
  const isNode =
    typeof window === 'undefined' && typeof document === 'undefined';
  if (isNode) {
    const { JSDOM } = await import('jsdom');
    return new JSDOM(html).window.document;
  } else {
    return Promise.resolve(new DOMParser().parseFromString(html, 'text/html'));
  }
};
