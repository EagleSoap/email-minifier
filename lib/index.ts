export { EmailMinifier } from './email-minifier';
