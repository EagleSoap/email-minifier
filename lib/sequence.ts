/**
 * Generates a sequence of alphabetical characters.
 */
export class Sequence {
  private current: string[];

  constructor() {
    this.current = ['a'];
  }

  next() {
    const sequence = this.current.join('');
    let i = this.current.length - 1;

    while (i >= 0) {
      if (this.current[i] !== 'z') {
        this.current[i] = this.getNextChar(this.current[i]);
        break;
      } else {
        this.current[i] = 'a';
        i--;
      }
    }

    if (i < 0) {
      this.current.unshift('a');
    }

    return sequence;
  }

  private getNextChar(char: string): string {
    const charCode = char.charCodeAt(0);
    const nextCharCode = charCode + 1;
    return String.fromCharCode(nextCharCode);
  }
}
