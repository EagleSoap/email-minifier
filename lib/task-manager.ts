import { Context, Task, TaskType } from './types';
import { minify as minifyCSS } from 'csso';

export class TaskManager {
  private mapping: Record<
    TaskType,
    (
      el: Element,
      context: Context,
      // Feel free to submit PRs to avoid using any here
      options?: any,
    ) => Task[]
  >;
  private tasks: Task[];

  constructor() {
    this.tasks = [];
    this.mapping = {
      'minify-ids': (el: Element, context: Context) => {
        if (!el.id) {
          return [];
        }
        const id = el.id;
        const {
          styles,
          idSequence,
          mapping: { ids },
        } = context;
        if (new RegExp(`#${id}`).test(styles)) {
          let newId = '';
          if (ids.has(id)) {
            newId = ids.get(id) as string;
          } else {
            newId = idSequence.next();
            ids.set(id, newId);
          }
          return [{ el, action: 'set-attribute', name: 'id', value: newId }];
        } else {
          return [{ el, action: 'remove-attribute', name: 'id' }];
        }
      },
      'minify-classes': (el: Element, context: Context) => {
        const classNames = Array.from(el.classList.values());
        const {
          styles,
          classSequence,
          mapping: { classes },
        } = context;
        if (
          classNames.length > 0 &&
          classNames.some((className) =>
            new RegExp(`.${className}`).test(styles),
          )
        ) {
          const newClassNames: string[] = [];
          classNames.forEach((className) => {
            let newClassName = '';
            if (classes.has(className)) {
              newClassName = classes.get(className) as string;
            } else {
              newClassName = classSequence.next();
              classes.set(className, newClassName);
            }
            newClassNames.push(newClassName);
          });
          return [
            {
              el,
              action: 'set-attribute',
              name: 'class',
              value: newClassNames.join(' '),
            },
          ];
        } else {
          return [{ el, action: 'remove-attribute', name: 'class' }];
        }
      },
      'remove-unused-attrs': (
        el: Element,
        context: Context,
        { matches } = { matches: [] },
      ) => {
        const tasks: Task[] = [];
        const { styles } = context;
        if (matches.length === 0) return [];

        el.getAttributeNames()
          .filter((attr) => {
            return matches.some(
              (reg: RegExp) => reg.test(attr) && !styles.includes(`[${attr}`), // like span[data-anything]
            );
          })
          .forEach((attr) => {
            tasks.push({
              el,
              action: 'remove-attribute',
              name: attr,
            });
          });
        return tasks;
      },
      // Dataset attrs are customized, we can safely minify them.
      'minify-dataset-attrs': (el: Element, context: Context) => {
        const tasks: Task[] = [];
        const {
          mapping: { dataSets },
        } = context;
        el.getAttributeNames()
          .filter((attr) => attr.startsWith('data-'))
          .forEach((attr) => {
            const newAttribute = attr
              .split('-')
              .map((x) => x[0])
              .flat()
              .join(''); // data-offset-key -> dok
            tasks.push({ el, action: 'remove-attribute', name: attr });
            tasks.push({
              el,
              action: 'set-attribute',
              name: newAttribute,
              value: '',
            });
            dataSets.set(attr, newAttribute);
          });
        return tasks;
      },
      'minify-styles': (
        el: Element,
        context: Context,
        {
          minifyStyles = true,
        }: {
          minifyStyles: boolean;
        },
      ) => {
        const tasks: Task[] = [];
        const {
          mapping: { ids, classes, dataSets },
        } = context;

        // Replace id selectors. eg: #id
        for (const key of Array.from(ids.keys()).sort(
          (a, b) => b.length - a.length,
        )) {
          tasks.push({
            el,
            action: 'minify-css-selector',
            originSelector: `#${key}`,
            newSelector: `#${ids.get(key)}`,
          });
        }

        // Replace class selectors. eg: .class
        for (const key of Array.from(classes.keys()).sort(
          (a, b) => b.length - a.length,
        )) {
          tasks.push({
            el,
            action: 'minify-css-selector',
            originSelector: `.${key}`,
            newSelector: `.${classes.get(key)}`,
          });
        }

        // Replace dataset property selectors. eg: span[data-offset-key]
        for (const key of Array.from(dataSets.keys()).sort(
          (a, b) => b.length - a.length,
        )) {
          tasks.push({
            el,
            action: 'minify-css-selector',
            originSelector: `[${key}`,
            newSelector: `[${dataSets.get(key)}`,
          });
        }
        minifyStyles && tasks.push({ el, action: 'minify-style' });
        return tasks;
      },
    };
  }

  add(
    type: TaskType,
    el: Element,
    context: Context,
    // Feel free to submit PRs to avoid using any here
    options?: any,
  ) {
    this.tasks.push(...this.mapping[type](el, context, options));
    return this;
  }

  runAll = () => {
    const runner = (task: Task) => {
      const { el } = task;
      switch (task.action) {
        case 'set-attribute':
          el.setAttribute(task.name, task.value);
          break;
        case 'remove-attribute':
          el.removeAttribute(task.name);
          break;
        case 'minify-css-selector':
          if (el.textContent) {
            el.textContent = el.textContent.replaceAll(
              task.originSelector,
              task.newSelector,
            );
          }
          break;
        case 'minify-style':
          if (el.textContent) {
            el.textContent = minifyCSS(el.textContent).css;
          }
          break;
      }
    };
    this.tasks.forEach(runner);
    const ranTasks = [...this.tasks];
    this.tasks = [];
    return ranTasks;
  };
}
