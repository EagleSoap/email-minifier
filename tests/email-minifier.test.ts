import fs from 'fs';
import path from 'path';

import { EmailMinifier } from '../lib/email-minifier';

const getTestData = (filename: string) => {
  return fs.readFileSync(path.resolve(__dirname, 'fixtures', filename), {
    encoding: 'utf-8',
  });
};

/**
 * Create an email body with the standard html format
 */
const createEmailBody = (content: string, css: string[] = []) => {
  return `
  <!DOCTYPE html>
  <html xmlns="http://www.w3.org/1999/xhtml">
  <head>
      ${css?.map((c) => `<style>${c}</style>`).join('\n')}
  </head>
  <body>
      ${content}
  </body>
  </html>`;
};

const removeUnusedAttrs = [
  'data-*',
  'autocorrect',
  'spellcheck',
  'contenteditable',
].map((x) => new RegExp(`^${x}`));

describe('email-minifier', () => {
  describe('Tasks', () => {
    it('should not generate tasks if incorrect email body provided', async () => {
      const emailBody = 'invalid email body';
      const minifier = new EmailMinifier(emailBody);
      const { tasks } = await minifier.minify();
      expect(tasks.length).toEqual(0);
    });

    it('should generate minify id tasks', async () => {
      const emailBody = createEmailBody(
        `<div id="test-id-1" ></div>
          <div id="test-id-2" ></div>
          <div id="test-id-unused" id-will-be-remove ></div>`,
        [`#test-id-1, #test-id-2 { font-size: 16px; }`],
      );
      const minifier = new EmailMinifier(emailBody);
      const { tasks } = await minifier.minify();

      const removeIdTasks = tasks.filter(
        (x) => x.action === 'remove-attribute' && x.name === 'id',
      );
      expect(removeIdTasks).toHaveLength(1);
      const [removeIdTask] = removeIdTasks;
      expect(removeIdTask.el.getAttribute('id-will-be-remove')).toEqual('');
      expect(removeIdTask.el.id).toEqual('');

      const setIdTasks = tasks.filter(
        (x) => x.action === 'set-attribute' && x.name === 'id',
      );
      expect(setIdTasks).toHaveLength(2);

      expect(setIdTasks).toContainEqual({
        el: expect.any(Object),
        action: 'set-attribute',
        name: 'id',
        value: 'a',
      });
      expect(setIdTasks).toContainEqual({
        el: expect.any(Object),
        action: 'set-attribute',
        name: 'id',
        value: 'b',
      });
    });

    it('should generate minify class tasks', async () => {
      const emailBody = createEmailBody(
        `<div class="test-class-reused" ></div>
          <div class="test-class-reused test-class-1" ></div>
          <div class="test-class-unused" class-will-be-removed ></div>`,
        [`.test-class-reused, .test-class-1 { font-size: 16px; }`],
      );
      const minifier = new EmailMinifier(emailBody);

      const { tasks } = await minifier.minify();
      const removeClassTasks = tasks.filter(
        (x) => x.action === 'remove-attribute' && x.name === 'class',
      );
      expect(removeClassTasks).toHaveLength(1);
      const [removeIdTask] = removeClassTasks;
      expect(removeIdTask.el.getAttribute('class-will-be-removed')).toEqual('');
      expect(removeIdTask.el.className).toEqual('');

      const setClassTasks = tasks.filter(
        (x) => x.action === 'set-attribute' && x.name === 'class',
      );
      expect(setClassTasks).toHaveLength(2);

      expect(setClassTasks).toContainEqual({
        el: expect.any(Object),
        action: 'set-attribute',
        name: 'class',
        value: 'a',
      });
      expect(setClassTasks).toContainEqual({
        el: expect.any(Object),
        action: 'set-attribute',
        name: 'class',
        value: 'a b',
      });
    });

    it('should generate minify style tasks', async () => {
      const emailBody = createEmailBody(
        `
        <div class="test-class-1"></div>
        <div class="test-class-2"></div>`,
        [
          `.test-class-1 {font-size: 16px; }`,
          `.test-class-2 {font-size: 16px; }`,
        ],
      );
      const minifier = new EmailMinifier(emailBody);

      const { tasks } = await minifier.minify();
      const minifyCSSTasks = tasks.filter((x) => x.action === 'minify-style');
      expect(minifyCSSTasks).toHaveLength(2);
    });

    it('should generate remove unused attribute tasks', async () => {
      const emailBody = createEmailBody(
        `<div contenteditable spellcheck autocorrect data-test-id="data-test-id" data-unremoved="data-unremoved" ></div>`,
      );
      const minifier = new EmailMinifier(emailBody);

      const { tasks } = await minifier.minify({
        removeUnusedAttrs,
      });
      expect(tasks).toContainEqual({
        el: expect.any(Object),
        action: 'remove-attribute',
        name: 'contenteditable',
      });
      expect(tasks).toContainEqual({
        el: expect.any(Object),
        action: 'remove-attribute',
        name: 'data-test-id',
      });
      expect(tasks).toContainEqual({
        el: expect.any(Object),
        action: 'remove-attribute',
        name: 'spellcheck',
      });
      expect(tasks).toContainEqual({
        el: expect.any(Object),
        action: 'remove-attribute',
        name: 'autocorrect',
      });
    });

    it('should generate minify data-* attribute tasks', async () => {
      const emailBody = createEmailBody(
        `<div data-test-name="data-test-name" data-test-age="data-test-age" ></div>`,
      );
      const minifier = new EmailMinifier(emailBody);

      const { tasks } = await minifier.minify();
      expect(tasks).toContainEqual({
        el: expect.any(Object),
        action: 'remove-attribute',
        name: 'data-test-name',
      });
      expect(tasks).toContainEqual({
        el: expect.any(Object),
        action: 'set-attribute',
        name: 'dtn',
        value: '',
      });
      expect(tasks).toContainEqual({
        el: expect.any(Object),
        action: 'remove-attribute',
        name: 'data-test-age',
      });
      expect(tasks).toContainEqual({
        el: expect.any(Object),
        action: 'set-attribute',
        name: 'dta',
        value: '',
      });
    });
  });

  describe('Options', () => {
    it('should generate tasks follow options.minifyIds', async () => {
      const emailBody = getTestData('full-featured-email-body');
      const minifier = new EmailMinifier(emailBody);
      expect((await minifier.minify({ minifyIds: true })).tasks).toContainEqual(
        {
          action: 'set-attribute',
          name: 'id',
          el: expect.any(Object),
          value: expect.any(String),
        },
      );
      expect(
        (await minifier.minify({ minifyIds: false })).tasks,
      ).not.toContainEqual({
        action: 'set-attribute',
        name: 'id',
        el: expect.any(Object),
        value: expect.any(String),
      });
    });

    it('should generate tasks follow options.minifyClasses', async () => {
      const emailBody = getTestData('full-featured-email-body');
      const minifier = new EmailMinifier(emailBody);
      expect(
        (await minifier.minify({ minifyClasses: true })).tasks,
      ).toContainEqual({
        action: 'set-attribute',
        name: 'class',
        el: expect.any(Object),
        value: expect.any(String),
      });
      expect(
        (await minifier.minify({ minifyClasses: false })).tasks,
      ).not.toContainEqual({
        action: 'set-attribute',
        name: 'class',
        el: expect.any(Object),
        value: expect.any(String),
      });
    });

    it('should generate tasks follow options.minifyStyles', async () => {
      const emailBody = getTestData('full-featured-email-body');
      const minifier = new EmailMinifier(emailBody);
      expect(
        (await minifier.minify({ minifyStyles: true })).tasks,
      ).toContainEqual({
        action: 'minify-style',
        el: expect.any(Object),
      });
      expect(
        (await minifier.minify({ minifyStyles: false })).tasks,
      ).not.toContainEqual({
        action: 'minify-style',
        el: expect.any(Object),
      });
    });

    it('should generate tasks follow options.minifyDatasets', async () => {
      const emailBody = getTestData('full-featured-email-body');
      const minifier = new EmailMinifier(emailBody);
      expect(
        (await minifier.minify({ minifyDatasets: true })).tasks,
      ).toContainEqual({
        action: 'remove-attribute',
        name: expect.stringMatching(/^data/),
        el: expect.any(Object),
      });
      expect(
        (await minifier.minify({ minifyDatasets: false })).tasks,
      ).not.toContainEqual({
        action: 'remove-attribute',
        name: expect.stringMatching(/^data/),
        el: expect.any(Object),
      });
    });

    it('should generate tasks follow options.removeUnusedAttrs', async () => {
      const emailBody = getTestData('full-featured-email-body');
      const minifier = new EmailMinifier(emailBody);
      expect(
        (await minifier.minify({ removeUnusedAttrs })).tasks,
      ).toContainEqual({
        action: 'remove-attribute',
        name: expect.stringMatching(/contenteditable/),
        el: expect.any(Object),
      });
      expect(
        (await minifier.minify({ removeUnusedAttrs: false })).tasks,
      ).not.toContainEqual({
        action: 'remove-attribute',
        name: expect.stringMatching(/contenteditable/),
        el: expect.any(Object),
      });
    });
  });

  describe('Snapshots', () => {
    it('should minify email correctly', async () => {
      const emailBody = getTestData('full-featured-email-body');
      const minifier = new EmailMinifier(emailBody);
      const { original, minified } = await minifier.minify({
        removeUnusedAttrs,
      });
      expect(minified!.length).toBeLessThan(original.length);
      expect(minified).toMatchSnapshot();
    });
  });
});
