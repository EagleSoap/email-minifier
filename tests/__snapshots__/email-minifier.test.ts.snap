// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`email-minifier Snapshots should minify email correctly 1`] = `
"<html xmlns="http://www.w3.org/1999/xhtml"><head>
    <style>.a{font-size:16px}.b{font-size:16px!important}#a,.c,.d,div[dt=""],div[dti]{font-size:16px}@media only screen and (min-width:481px){.d{font-size:16px}}@media (prefers-color-scheme:dark){.d{font-size:16px}}</style>
  </head>
  <body>
    <div>
      My class will be removed
    </div>
    <div class="a">
      My class will be minified as a single char
    </div>
    <div class="b c">My Class will be minified as two single chars</div>
    <div class="c d">
      My first class will be reuse, and second one will be minified as a
      single char
      
    </div>
    <div id="a">
      My id will be minified as a single char
    </div>
    <div dt="" dti="">
      My first dataset attribute will be minified as \`dt=""\` and the second
      one will be minified as \`dti=""\`
    </div>
    <div>My contenteditable attribute will be removed</div>
  

</body></html>"
`;
